apps/remote_control/lib/lexical/remote_control/code_intelligence/definition.ex

```diff
@@ -0,0 +1,58 @@
+defmodule Lexical.RemoteControl.CodeIntelligence.Definition do
+  alias Lexical.SourceFile
+  alias Lexical.SourceFile.Position
+  alias Lexical.SourceFile.Range
+  alias Lexical.Text
+
+  require Logger
+
+  @spec definition(SourceFile.t(), Position.t()) ::
+          {:ok, {SourceFile.t(), Range.t()} | nil} | {:error, String.t()}
+  def definition(%SourceFile{} = source_file, %Position{} = position) do
+    maybe_location =
+      source_file
+      |> SourceFile.to_string()
+      |> ElixirSense.definition(position.line, position.character)
+
+    parse_location(maybe_location, source_file)
+  end
+
+  defp parse_location(%ElixirSense.Location{} = location, source_file) do
+    %{file: file, line: line, column: column} = location
+    file_path = file || source_file.path
+    uri = SourceFile.Path.ensure_uri(file_path)
+
+    with {:ok, source_file} <- SourceFile.Store.open_temporary(uri),
+         {:ok, text} <- SourceFile.fetch_text_at(source_file, line) do
+      {:ok, {source_file, to_precise_range(text, line, column)}}
+    else
+      _ ->
+        {:error, "Could not open source file: #{inspect(file_path)}"}
```

Comment:
this error message could possibly be wrong; you could also be fetching a line that's out of bounds.